From 185cc4200ed83637d8515efe7ebd55c3831c6a15 Mon Sep 17 00:00:00 2001
From: Aleksey Orekhovsky <aaa4xu@gmail.com>
Date: Fri, 22 Aug 2025 14:24:38 +0700
Subject: [PATCH] Add tracking of factory construction

---
 src/core/game/UnitImpl.ts | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/core/game/UnitImpl.ts b/src/core/game/UnitImpl.ts
index 899e56d7..bf9f3be9 100644
--- a/src/core/game/UnitImpl.ts
+++ b/src/core/game/UnitImpl.ts
@@ -73,6 +73,7 @@ export class UnitImpl implements Unit {
       case UnitType.DefensePost:
       case UnitType.SAMLauncher:
       case UnitType.City:
+      case UnitType.Factory:
         this.mg.stats().unitBuild(_owner, this._type);
     }
   }
-- 
2.39.5 (Apple Git-154)

