diff --git a/src/ReplayPlaybackEngine.ts b/src/ReplayPlaybackEngine.ts
index 0c38e697..1b5d5b68 100644
--- a/src/ReplayPlaybackEngine.ts
+++ b/src/ReplayPlaybackEngine.ts
@@ -41,9 +41,6 @@ export class ReplayPlaybackEngine implements PlaybackEngine<GameRecord> {
     const runner = await this.createGameRunner(record.info);
 
     const winner = record.info.winner;
-    if (!winner) {
-      throw new Error("No winner found in replay");
-    }
 
     const spawns = new Map<string, [number, number]>();
     const turns: Array<Record<string, LensStatsPlayerTurn>> = [];
