diff --git a/src/core/execution/ExecutionManager.ts b/src/core/execution/ExecutionManager.ts
index 192c0f9a..18930d53 100644
--- a/src/core/execution/ExecutionManager.ts
+++ b/src/core/execution/ExecutionManager.ts
@@ -43,7 +43,6 @@ export class Executor {
   createExec(intent: Intent): Execution {
     const player = this.mg.playerByClientID(intent.clientID);
     if (!player) {
-      console.warn(`player with clientID ${intent.clientID} not found`);
       return new NoOpExecution();
     }
     const playerID = player.id();
diff --git a/src/core/execution/WarshipExecution.ts b/src/core/execution/WarshipExecution.ts
index 4992b8d0..2403840f 100644
--- a/src/core/execution/WarshipExecution.ts
+++ b/src/core/execution/WarshipExecution.ts
@@ -222,7 +222,6 @@ export class WarshipExecution implements Execution {
         this.warship.touch();
         return;
       case PathFindResultType.PathNotFound:
-        consolex.warn(`path not found to target tile`);
         this.warship.setTargetTile(undefined);
         break;
     }
